File: tests/nn_xor.py
from cnn import Sequential, Dense, Input, Sigmoid, ReLU
import numpy as np

relu = ReLU()
sigmoid = Sigmoid()
model = Sequential(0.5)

inpt = Input(2)
d1 = Dense(2, relu)
d2 = Dense(2, relu)
d3 = Dense(1, sigmoid)

model.add_layer(inpt)
model.add_layer(d1)
model.add_layer(d2)
model.add_layer(d3)

x_train = np.array([[0, 0], [0, 1], [1, 0], [1, 1]])
y_train = np.array([0, 1, 1, 0])

model.run(x_train, y_train, 5000, 1)
model.save("test_xor.cnn")
print(x_train)
print(y_train)
print("Prediction:")
print(model.predict(x_train))

new_model = Sequential(0.5)
new_model.load("test_xor.cnn")
print("Prediction (loaded):")
print(new_model.predict(x_train))

File: cpp/example/nn_xor.cpp
#include "../include/cnn.hpp"
#include "../include/cnn/Utils.hpp"

#include <fstream>
#include <iostream>

using namespace std;
int main()
{
    Sigmoid sig = Sigmoid();
    ReLU relu = ReLU();
    Sequential *model = new Sequential(0.5);

    model->addLayer(new Input(2));
    model->addLayer(new Dense(2, &relu));
    model->addLayer(new Dense(2, &relu));
    model->addLayer(new Dense(1, &sig));

    Matrix x_train = Matrix::Zero(4, 2);
    Matrix y_train = Matrix::Zero(4, 1);

    x_train(1, 1) = 1;
    x_train(2, 0) = 1;
    x_train(3, 0) = 1;
    x_train(3, 1) = 1;
    y_train(1, 0) = 1;
    y_train(2, 0) = 1;

    model->run(x_train, y_train, 5000, 1);
    model->saveModel("test_model.cnn");

    cout << x_train << endl;
    cout << y_train << endl;
    cout << "Prediction:" << endl;
    cout << model->predict(x_train) << endl;
    Model *new_model = new Sequential(0.2);
    new_model->loadModel("test_model.cnn");
    cout << "Prediction (loaded):" << endl;
    cout << new_model->predict(x_train) << endl;

    return 0;
}